app/src/main/java/com/nutomic/syncthingandroid/util/ConfigXml.java

```diff
@@ -342,14 +356,34 @@ private void saveChanges() {
         Log.i(TAG, "Writing updated config file");
         File mConfigTempFile = Constants.getConfigTempFile(mContext);
         try {
+            // Write XML header.
+            FileOutputStream fileOutputStream = new FileOutputStream(mConfigTempFile);
+            fileOutputStream.write("<?xml version=\"1.0\" encoding=\"UTF-8\"?>".getBytes("UTF-8"));
+
+            // Prepare Object-to-XML transform.
             TransformerFactory transformerFactory = TransformerFactory.newInstance();
             Transformer transformer = transformerFactory.newTransformer();
-            DOMSource domSource = new DOMSource(mConfig);
-            StreamResult streamResult = new StreamResult(mConfigTempFile);
-            transformer.transform(domSource, streamResult);
+            transformer.setOutputProperty(OutputKeys.METHOD, "xml");
+            transformer.setOutputProperty(OutputKeys.ENCODING, "UTF-16");
+            transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes");
```

Comment:
Does thos option behave incorrectly hence you need to write the header manually?

Reply:
Yes, that's correct.

Reply:
What part is missing in the header with that option?

Reply:
If you don't write the header manually and change the encoding syncthing cannot start. Seems to be a bug in the xml library.

Reply:
Sure. I am just asking how the header differs between what we are writing and what the library writes, so I could understand what the problem is.

Potentially it could also be explained in a comment above for any future reader.

Reply:
The library still would write UTF-8 so parsing UTF-16 stuff would throw an error.

Reply:
I will defer to @imsodin, @nutomic and @imsodin for approval, as I simply don't understand how:

1. Reading as UTF-8
2. Writing a header as UTF-8
3. Writing data as supposedly UTF-16 (yet lying in the header of the enconding being utf-8 which I presume is default anyway)
4. Setting indentation to no which however supposedly formats the data correctly

fixes this issue. 

I don't dispute it works, but I want to understand why settings are what they are and why this works because of them.

If others feel they understand this as it is, or if they don't think answers to this are necessery they can feel free to approve and merge.

As a minimum I feel this could be simplified (using built in header writer as I think it can be mended via options to set the right output, setting the encoding output to utf-8, setting indentation to true)

Reply:
I agree with Audrius, this code looks weird and overly complicated. But I havent done Android development in over a year, and I dont have the time to dig into this myself. 

That said, I wouldnt be completely opposed to merging it, if it works and you can add some comments explaining why you do this. Maybe with links to stackoverflow or wherever you found the solution.

Perhaps there is also a library that handles this better?